datafusion/physical-expr/src/simplifier/unwrap_cast.rs

```diff
@@ -0,0 +1,621 @@
+    literal_value: &ScalarValue,
+    op: Operator,
+    schema: &Schema,
+) -> Result<Option<Arc<dyn PhysicalExpr>>> {
+    // Get the data type of the inner expression
+    let inner_type = inner_expr.data_type(schema)?;
+
+    // Try to cast the literal to the inner expression's type
+    if let Some(casted_literal) = try_cast_literal_to_type(literal_value, &inner_type) {
+        let literal_expr = lit(casted_literal);
+        let binary_expr = BinaryExpr::new(inner_expr, op, literal_expr);
+        return Ok(Some(Arc::new(binary_expr)));
+    }
+
+    Ok(None)
+}
+
+/// Swap comparison operators for right-side cast unwrapping
+fn swap_operator(op: Operator) -> Operator {
+    match op {
+        Operator::Lt => Operator::Gt,
+        Operator::LtEq => Operator::GtEq,
+        Operator::Gt => Operator::Lt,
+        Operator::GtEq => Operator::LtEq,
+        // Symmetric operators remain the same
+        Operator::Eq | Operator::NotEq => op,
+        // Other operators are not swappable
+        _ => op,
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::expressions::{col, lit};
+    use arrow::datatypes::{DataType, Field, Schema};
+    use datafusion_common::ScalarValue;
+    use datafusion_expr::Operator;
+
+    /// Check if an expression is a cast expression
+    fn is_cast_expr(expr: &Arc<dyn PhysicalExpr>) -> bool {
+        expr.as_any().downcast_ref::<CastExpr>().is_some()
+            || expr.as_any().downcast_ref::<TryCastExpr>().is_some()
+    }
+
+    /// Check if a binary expression is suitable for cast unwrapping
+    fn is_binary_expr_with_cast_and_literal(binary: &BinaryExpr) -> bool {
+        // Check if left is cast and right is literal
+        let left_cast_right_literal = is_cast_expr(binary.left())
+            && binary.right().as_any().downcast_ref::<Literal>().is_some();
+
+        // Check if left is literal and right is cast
+        let left_literal_right_cast =
+            binary.left().as_any().downcast_ref::<Literal>().is_some()
+                && is_cast_expr(binary.right());
+
+        left_cast_right_literal || left_literal_right_cast
+    }
+
+    fn test_schema() -> Schema {
+        Schema::new(vec![
+            Field::new("c1", DataType::Int32, false),
+            Field::new("c2", DataType::Int64, false),
+            Field::new("c3", DataType::Utf8, false),
+        ])
+    }
+
+    #[test]
+    fn test_unwrap_cast_in_binary_comparison() {
+        let schema = test_schema();
+
+        // Create: cast(c1 as INT64) > INT64(10)
+        let column_expr = col("c1", &schema).unwrap();
+        let cast_expr = Arc::new(CastExpr::new(column_expr, DataType::Int64, None));
+        let literal_expr = lit(ScalarValue::Int64(Some(10)));
+        let binary_expr =
+            Arc::new(BinaryExpr::new(cast_expr, Operator::Gt, literal_expr));
+
+        // Apply unwrap cast optimization
+        let result = unwrap_cast_in_comparison(binary_expr, &schema).unwrap();
```

Comment:
As a a minor comment it would be nice if we could reduce the boiler plate in these tests -- specifically, you could perhaps factor out the unwrapping and verifying that it was transformed